opm/simulators/linalg/cuistl/CuJac.cpp

```diff
@@ -0,0 +1,182 @@
+{
+}
+
+template <class M, class X, class Y, int l>
+void
+CuJac<M, X, Y, l>::apply(X& x, const Y& b)
+{
+    // x_{n+1} = x_n + w * (D^-1 * (b - Ax_n) )
+    // cusparseDbsrmv computes -Ax_n + b
+    // cuda kernel computes D^-1 * (b- Ax_n), call this rhs
+    // use an axpy to compute x + w*rhs
+      
+    const field_type one = 1.0, neg_one = -1.0;
+
+    const auto numberOfRows = detail::to_int(m.N());
+    const auto numberOfColumns = detail::to_int(m.N());
+    const auto numberOfNonzeroBlocks = detail::to_int(m.nonzeroes());
+    const auto blockSize = detail::to_int(m.blockSize());
+
+    auto nonZeroValues = m.getNonZeroValues().data();
+    auto rowIndices = m.getRowIndices().data();
+    auto columnIndices = m.getColumnIndices().data();
+
+    // TODO: avoid making this copy, currently forced to because parameter is a const
+    d_resultBuffer = b; 
+
+    // bsrmv computes b - Ax
+    OPM_CUSPARSE_SAFE_CALL(detail::cusparseBsrmv(m_cuSparseHandle.get(),
+                                                 detail::CUSPARSE_MATRIX_ORDER,
+                                                 CUSPARSE_OPERATION_NON_TRANSPOSE,
+                                                 numberOfRows,
+                                                 numberOfColumns,
+                                                 numberOfNonzeroBlocks,
+                                                 &neg_one,
+                                                 m_description->get(),
+                                                 nonZeroValues,
+                                                 rowIndices,
+                                                 columnIndices,
+                                                 blockSize,
+                                                 x.data(),
+                                                 &one,
+                                                 d_resultBuffer.data()));
+
+    // multiply D^-1 with (b-Ax)
+    detail::blockVectorMultiplicationAtAllIndices(m_diagInvFlattened.data(), detail::to_size_t(numberOfRows), detail::to_size_t(blockSize), d_resultBuffer.data());
+
+    // Finish adding w*rhs to x
+    OPM_CUBLAS_SAFE_CALL(detail::cublasAxpy(m_cuBlasHandle.get(),
+                                            numberOfRows*blockSize,
+                                            &m_w,
+                                            d_resultBuffer.data(),
+                                            1,
+                                            x.data(),
+                                            1));
+}
+
+template <class M, class X, class Y, int l>
+void
+CuJac<M, X, Y, l>::post([[maybe_unused]] X& x)
+{
+}
+
+template <class M, class X, class Y, int l>
+Dune::SolverCategory::Category
+CuJac<M, X, Y, l>::category() const
+{
+    return Dune::SolverCategory::sequential;
+}
+
+template <class M, class X, class Y, int l>
+void
+CuJac<M, X, Y, l>::update()
+{
+    //TODO: this update of m is done twice when it is initialized
```

Comment:
Suggest that you instead of creating `m` in the initializer list of the constructor build it here in update(), with an `if` to distinguish between the "need to create from scratch" and "only need to update nonzero values" cases.

Reply:
In this case the update is 2 lines and I only need one of them to be added to the constructor, so I will instead just omit the update() in the constructor and simply flatten the diagonal immediately. I think that will be clean.